database/pgsql/lock.go

```diff
@@ -19,71 +19,77 @@ import (
 	"time"
 
 	log "github.com/sirupsen/logrus"
-
-	"github.com/coreos/clair/pkg/commonerr"
 )
 
 const (
-	soiLock           = `INSERT INTO lock(name, owner, until) VALUES ($1, $2, $3)`
 	searchLock        = `SELECT until FROM Lock WHERE name = $1`
 	updateLock        = `UPDATE Lock SET until = $3 WHERE name = $1 AND owner = $2`
 	removeLock        = `DELETE FROM Lock WHERE name = $1 AND owner = $2`
 	removeLockExpired = `DELETE FROM LOCK WHERE until < CURRENT_TIMESTAMP`
+
+	soiLock = `WITH new_lock AS (
+		INSERT INTO lock (name, owner, until) 
+		          VALUES (  $1,    $2,    $3)
+		          WHERE NOT EXISTS (SELECT id FROM lock WHERE name = $1)
+		          RETURNING owner, until
+	)
+	SELECT * FROM new_lock
+	UNION
+	SELECT owner, until FROM lock WHERE name = $1`
 )
 
 var (
 	errLockNotFound = errors.New("lock is not in database")
 )
 
-// Lock tries to set a temporary lock in the database.
-//
-// Lock does not block, instead, it returns true and its expiration time
-// is the lock has been successfully acquired or false otherwise.
-func (tx *pgSession) Lock(name string, owner string, duration time.Duration, renew bool) (bool, time.Time, error) {
-	if name == "" || owner == "" || duration == 0 {
-		log.Warning("could not create an invalid lock")
-		return false, time.Time{}, commonerr.NewBadRequestError("Invalid Lock Parameters")
+func (tx *pgSession) AcquireLock(lockName, whoami string, desiredDuration time.Duration) (bool, time.Time, error) {
+	if lockName == "" || whoami == "" || desiredDuration == 0 {
+		panic("invalid lock parameters")
 	}
 
-	until := time.Now().Add(duration)
-	if renew {
-		defer observeQueryTime("Lock", "update", time.Now())
-		// Renew lock.
-		r, err := tx.Exec(updateLock, name, owner, until)
-		if err != nil {
-			return false, until, handleError("updateLock", err)
-		}
-
-		if n, err := r.RowsAffected(); err == nil {
-			return n > 0, until, nil
-		}
-		return false, until, handleError("updateLock", err)
-	} else if err := tx.pruneLocks(); err != nil {
-		return false, until, err
+	if err := tx.pruneLocks(); err != nil {
+		return false, time.Time{}, err
 	}
 
-	// Lock.
+	var (
+		desiredLockedUntil = time.Now().Add(desiredDuration)
+
+		lockedUntil time.Time
+		lockOwner   string
+	)
+
 	defer observeQueryTime("Lock", "soiLock", time.Now())
-	_, err := tx.Exec(soiLock, name, owner, until)
+	err := tx.QueryRow(soiLock, lockName, whoami, desiredLockedUntil).Scan(&lockOwner, lockedUntil)
```

Comment:
do we need `&lockedUntil`?

Reply:
In the case where you don't get the lock, you want the time it'll expire, so you can try and grab it again around then.